### ISSUE TITLE
Custom backbone for multiple inputs

### ISSUE THREAD

**Issue by bw4sz** [weecology/DeepForest_demos#3]
# Goal
To create a short and clear ipython notebook outlining how to manipulate the retinanet backbone of the DeepForest model for multiple inputs.

## Rationale

For several of the upcoming projects we will need to move away from the traditional architecture and develop custom models. Applications include:

* Multi-temporal inputs - RGB images from multiple years are used for joint prediction.
* Multi-sensor inputs - Any combination of LiDAR + RGB + Hyperspectral data for joint prediction.
* Metadata fusion - Adding image location data such as time/place as a feature alongside the imagery. This can be extended to ecological knowledge as well.
* Parameter reduction - Studying the relationship between speed and performance when minimizing model size.

## Current state of knowledge

DeepForest depends the tensorflow >2.0 implementation of keras-retinanet. A fork of the repo sits in the DeepForest source. Retinanets are a one shot model that use anchor boxes as region proposals. Retinanets have two main components, a backbone network that learns image features and an image pyramid network that scales these features and relates them to anchor box proposals. See:

https://towardsdatascience.com/review-retinanet-focal-loss-object-detection-38fba6afabe4
for a nice diagram.

Our goal is to only manipulate the backbone. Bonus points if we can maintain the imagenet pretraining weights as much as possible during this process. The current DeepForest model uses a resnet50 backbone, which come from keras_resnet. The main retinanet repo has implementations of a half-dozen other models, which will be useful in comparison of the model inputs. The source for the conversion from backbone to full retinanet is here:

https://github.com/fizyr/keras-retinanet/blob/4c131fbab33a83ea39ff2c2b31c0a6964a49e69f/keras_retinanet/models/retinanet.py#L238

On initial inspection, as long as the backbone has three pooling layers, it can be passed to retinanet:

https://github.com/fizyr/keras-retinanet/blob/4c131fbab33a83ea39ff2c2b31c0a6964a49e69f/keras_retinanet/models/retinanet.py#L127

## How to access a DeepForest backbone and look at current architecture.

```
from deepforest import deepforest
original = deepforest.deepforest()
#Load current weights
original.use_release()
original.model
print("The current model has {} layers".format(len(original.model.layers)))
```
The current model has 207 layers

```
[x.name for x in original.model.layers]
['input_1', 'padding_conv1', 'conv1', 'bn_conv1', 'conv1_relu', 'pool1', 'res2a_branch2a', 'bn2a_branch2a', 'res2a_branch2a_relu', 'padding2a_branch2b', 'res2a_branch2b', 'bn2a_branch2b', 'res2a_branch2b_relu', 'res2a_branch2c', 'res2a_branch1', 'bn2a_branch2c', 'bn2a_branch1', 'res2a', 'res2a_relu', 'res2b_branch2a', 'bn2b_branch2a', 'res2b_branch2a_relu', 'padding2b_branch2b', 'res2b_branch2b', 'bn2b_branch2b', 'res2b_branch2b_relu', 'res2b_branch2c', 'bn2b_branch2c', 'res2b', 'res2b_relu', 'res2c_branch2a', 'bn2c_branch2a', 'res2c_branch2a_relu', 'padding2c_branch2b', 'res2c_branch2b', 'bn2c_branch2b', 'res2c_branch2b_relu', 'res2c_branch2c', 'bn2c_branch2c', 'res2c', 'res2c_relu', 'res3a_branch2a', 'bn3a_branch2a', 'res3a_branch2a_relu', 'padding3a_branch2b', 'res3a_branch2b', 'bn3a_branch2b', 'res3a_branch2b_relu', 'res3a_branch2c', 'res3a_branch1', 'bn3a_branch2c', 'bn3a_branch1', 'res3a', 'res3a_relu', 'res3b_branch2a', 'bn3b_branch2a', 'res3b_branch2a_relu', 'padding3b_branch2b', 'res3b_branch2b', 'bn3b_branch2b', 'res3b_branch2b_relu', 'res3b_branch2c', 'bn3b_branch2c', 'res3b', 'res3b_relu', 'res3c_branch2a', 'bn3c_branch2a', 'res3c_branch2a_relu', 'padding3c_branch2b', 'res3c_branch2b', 'bn3c_branch2b', 'res3c_branch2b_relu', 'res3c_branch2c', 'bn3c_branch2c', 'res3c', 'res3c_relu', 'res3d_branch2a', 'bn3d_branch2a', 'res3d_branch2a_relu', 'padding3d_branch2b', 'res3d_branch2b', 'bn3d_branch2b', 'res3d_branch2b_relu', 'res3d_branch2c', 'bn3d_branch2c', 'res3d', 'res3d_relu', 'res4a_branch2a', 'bn4a_branch2a', 'res4a_branch2a_relu', 'padding4a_branch2b', 'res4a_branch2b', 'bn4a_branch2b', 'res4a_branch2b_relu', 'res4a_branch2c', 'res4a_branch1', 'bn4a_branch2c', 'bn4a_branch1', 'res4a', 'res4a_relu', 'res4b_branch2a', 'bn4b_branch2a', 'res4b_branch2a_relu', 'padding4b_branch2b', 'res4b_branch2b', 'bn4b_branch2b', 'res4b_branch2b_relu', 'res4b_branch2c', 'bn4b_branch2c', 'res4b', 'res4b_relu', 'res4c_branch2a', 'bn4c_branch2a', 'res4c_branch2a_relu', 'padding4c_branch2b', 'res4c_branch2b', 'bn4c_branch2b', 'res4c_branch2b_relu', 'res4c_branch2c', 'bn4c_branch2c', 'res4c', 'res4c_relu', 'res4d_branch2a', 'bn4d_branch2a', 'res4d_branch2a_relu', 'padding4d_branch2b', 'res4d_branch2b', 'bn4d_branch2b', 'res4d_branch2b_relu', 'res4d_branch2c', 'bn4d_branch2c', 'res4d', 'res4d_relu', 'res4e_branch2a', 'bn4e_branch2a', 'res4e_branch2a_relu', 'padding4e_branch2b', 'res4e_branch2b', 'bn4e_branch2b', 'res4e_branch2b_relu', 'res4e_branch2c', 'bn4e_branch2c', 'res4e', 'res4e_relu', 'res4f_branch2a', 'bn4f_branch2a', 'res4f_branch2a_relu', 'padding4f_branch2b', 'res4f_branch2b', 'bn4f_branch2b', 'res4f_branch2b_relu', 'res4f_branch2c', 'bn4f_branch2c', 'res4f', 'res4f_relu', 'res5a_branch2a', 'bn5a_branch2a', 'res5a_branch2a_relu', 'padding5a_branch2b', 'res5a_branch2b', 'bn5a_branch2b', 'res5a_branch2b_relu', 'res5a_branch2c', 'res5a_branch1', 'bn5a_branch2c', 'bn5a_branch1', 'res5a', 'res5a_relu', 'res5b_branch2a', 'bn5b_branch2a', 'res5b_branch2a_relu', 'padding5b_branch2b', 'res5b_branch2b', 'bn5b_branch2b', 'res5b_branch2b_relu', 'res5b_branch2c', 'bn5b_branch2c', 'res5b', 'res5b_relu', 'res5c_branch2a', 'bn5c_branch2a', 'res5c_branch2a_relu', 'padding5c_branch2b', 'res5c_branch2b', 'bn5c_branch2b', 'res5c_branch2b_relu', 'res5c_branch2c', 'bn5c_branch2c', 'res5c', 'res5c_relu', 'C5_reduced', 'P5_upsampled', 'C4_reduced', 'P4_merged', 'P4_upsampled', 'C3_reduced', 'P6', 'P3_merged', 'C6_relu', 'P3', 'P4', 'P5', 'P7', 'regression_submodel', 'classification_submodel', 'regression', 'classification']
```

Looking here: 
https://github.com/fizyr/keras-retinanet/blob/4c131fbab33a83ea39ff2c2b31c0a6964a49e69f/keras_retinanet/models/resnet.py#L112

The following layers are the ones actually passed to the retinanet backbone. 
```
import keras
import keras_resnet
import keras_resnet.models
inputs = keras.layers.Input(shape=(None, None, 3))
resnet = keras_resnet.models.ResNet50(inputs, include_top=False, freeze_bn=True)
backbone_layers=resnet.outputs[1:]
[x.name for x in backbone_layers]
```

```
['res3d_relu/Relu:0', 'res4f_relu/Relu:0', 'res5c_relu/Relu:0']
```

presumably that means that everything before these layers can be manipulated and most of the current code can be maintained? As an aside, I would love insight into why these layers were chosen for the feature pyramid network, I see several other relu activation layers (e.g. res5a_relu). That's a mystery to me.

## Measure of success
For an example network can the model learn and the loss drops using a toy dataset. @bw4sz still needs to make a toy dataset, but I think alot of this can start taking place.

## Notes

* I think it probably makes sense to begin by loading a current resnet architecture, manipulating the layers, and recompiling the model. If we feel like we've made progress we can do down the route of writing a model definition akin to the ones here:

https://github.com/fizyr/keras-retinanet/tree/master/keras_retinanet/models

* This is all leaving aside the semi-supervised pretraining
* There is a retinanet slack channel that is pretty active (kerasteam.slack.com #keras-retinanet)
* I'm not wholly sure how to point to a message thread on slack (on july 22nd, 2019), but there has been some discussion about VGG style backbones. To be clear I'm not tied to resnet, whatever we get to work is fine.

https://kerasteam.slack.com/archives/C7P6JEQDU/p1563779170011900
One of the devs writes

> "If I were you I would just use VGG as the example and look at the following line (https://github.com/fizyr/keras-retinanet/blob/master/keras_retinanet/models/vgg.py#L98). You just have to ensure that your backbone returns a similar list where the 1st block is 1/4th of the size of the input (so 2x 2x2 maxpooling for example), 2nd block is 1/8th of the size etc. Hope this makes a bit of sense.

@dstewart19 and I will try to work on this together, if willing.

**Comment by bw4sz** [comment]
toy dataset 

https://www.dropbox.com/sh/zkv5br22u037onx/AAADP-WV5Wm9c0vx_jBHK5U7a?dl=0
